hedera-node/hedera-app/src/main/java/com/hedera/node/app/blocks/impl/streaming/BlockBufferService.java

```
@@ -0,0 +1,604 @@
+                .getConfigData(BlockStreamConfig.class)
+                .blockBufferTtl();
+    }
+
+    /**
+     * @return the block period duration (i.e. the amount of time a single block represents)
+     */
+    private Duration blockPeriod() {
+        return configProvider
+                .getConfiguration()
+                .getConfigData(BlockStreamConfig.class)
+                .blockPeriod();
+    }
+
+    /**
+     * Sets the block node connection manager for notifications.
+     *
+     * @param blockNodeConnectionManager the block node connection manager
+     */
+    public void setBlockNodeConnectionManager(@NonNull final BlockNodeConnectionManager blockNodeConnectionManager) {
+        this.blockNodeConnectionManager =
+                requireNonNull(blockNodeConnectionManager, "blockNodeConnectionManager must not be null");
+    }
+
+    /**
+     * Opens a new block for streaming with the given block number. Creates a new BlockState, adds it to the buffer,
+     * and notifies block nodes if streaming is enabled. This will also attempt to prune older blocks from the buffer.
+     *
+     * @param blockNumber the block number
+     * @throws IllegalArgumentException if the block number is negative
+     */
+    public void openBlock(final long blockNumber) {
+        if (!isStreamingEnabled.get()) {
+            return;
+        }
+
+        if (blockNumber < 0) {
+            throw new IllegalArgumentException("Block number must be non-negative");
+        }
+
+        if (this.lastProducedBlockNumber >= blockNumber) {
+            logger.error(
+                    "Attempted to open a new block with number {}, but a block with the same or later number "
+                            + "(latest: {}) has already been opened",
+                    blockNumber,
+                    this.lastProducedBlockNumber);
+            throw new IllegalStateException("Attempted to open a new block with number " + blockNumber
```

If we throw here instead of just logging an `ERROR`, imo we must be 100% sure that the client code would never have a _survivable_ failure mode that repeats calls to `openBlock()`.

If there is a bug upstream that incorrectly, but non-fatally, repeats a call to `openBlock()`, throwing this exception would like turn a survivable failure mode into a fatal one. 😓

---

Personally, I would only ever fail to re-open a block _after_ the block proof is added, or perhaps after it is acknowledged (not totally sure which is more "correct").  That's the point where it is certain the block is closed and cannot be updated.

Note, we absolutely _can_ have a situation where block `N` must be appended after block `N+1` is opened, specifically we _can_ have a block proof for block `N` that is generated from the signatures on block `N+1` (or even later).

Just some thoughts.

---

Would this approach work?
- If an attempt is made to open a block that is not in the buffer and is older than the latest acknowledged block (i.e. trying to open a block that has been pruned), then it will be rejected.
- If an attempt is made to re-open a block that has it's block proof already sent (either pending an ack or already acked) then it will be rejected.
- If an attempt is made to re-open a block that has not been acked or had its block proof sent yet, then we allow it to be opened, we clear any items/requests associated with the previous version of the block, and restart streaming the block beginning with the header.

---

This should work.